### PR TITLE
[NIFI-9666] Separation of C2 Client(s) and the creation of the heartbeats

### PR DESCRIPTION
**Pull request by bejancsaba**
This PR is a proposal to separate as much logic as possible from NiFi framework core.
- The code was a little bit restructured and moved to c2-client-bundle. Also Heartbeat creation was extracted. 
- Runtime / dynamically changing parameters are provided via nifi framework core
- A new C2ClientService was introduced and the old NiFi specific one was renamed to C2NifiClientService while parts of it were moved so following the change inn the review could be more challenging because of this.

### CODE REVIEW

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/c2/C2NifiClientService.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+import java.util.Properties;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import static org.apache.commons.lang3.StringUtils.defaultIfBlank;
+
+public class C2NifiClientService {
+
+    private static final Logger logger = LoggerFactory.getLogger(C2NifiClientService.class);
+    private static final String minifiHome = System.getenv("MINIFI_HOME");
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/c2/C2NifiClientService.java]: I think we'll have to pass this "home" variable into the constructor for when we want to use NIFI_HOME or something else in the future

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/c2/C2NifiClientService.java [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    private static final String DEFAULT_CONF_DIR = "./conf";
+    private static final String DEFAULT_CONFIG_FILE = DEFAULT_CONF_DIR + "/bootstrap.conf";
+    private static final String ROOT_GROUP_ID = "root";
+    private static final String RUNTIME_MANIFEST_IDENTIFIER = "minifi";
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/c2/C2NifiClientService.java]: Same goes for the runtime identifier and type, we should provide it somehow if possible

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/c2/C2NifiClientService.java [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    private static final Integer TERMINATION_WAIT = 5000;
+
+    private final C2ClientService c2ClientService;
+    private final FlowController flowController;
+    private final ScheduledThreadPoolExecutor scheduledExecutorService = new ScheduledThreadPoolExecutor(1);
+    private final ExtensionManifestParser extensionManifestParser = new JAXBExtensionManifestParser();
+
+    final RuntimeManifestService runtimeManifestService =
+            new StandardRuntimeManifestService(ExtensionManagerHolder.getExtensionManager(), extensionManifestParser, RUNTIME_MANIFEST_IDENTIFIER, RUNTIME_TYPE);
+    private String bootstrapConfigFilename = System.getProperty("org.apache.nifi.minifi.bootstrap.config.file");
```
Comment (mattyb149) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/c2/C2NifiClientService.java]: Same here, anything MiNiFi-specific should somehow be configured on this client service if possible

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/c2/C2NifiClientService.java (hunk 1 above)
Comment (bejancsaba) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/c2/C2NifiClientService.java]: Restructured it as requested.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/c2/C2NifiClientService.java (hunk 2 above)
Comment (bejancsaba) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/c2/C2NifiClientService.java]: Restructured it as requested.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/c2/C2NifiClientService.java (hunk 3 above)
Comment (bejancsaba) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/c2/C2NifiClientService.java]: Restructured it as requested.

**Comment by bejancsaba** [comment]
@mattyb149  Thanks for the  review I addressed the comments. The idea is that everything is needed to configured for C2 comes via the properties so all that hardcoded logic is now externalised. For now to start minifi you will need to add additional properties to config.yml:

NiFi Properties Overrides:
  nifi.c2.enable: 'true'
  nifi.c2.config.directory: './conf'
  nifi.c2.runtime.manifest.identifier: 'minifi'
  nifi.c2.runtime.type: 'minifi-java'
  nifi.c2.rest.url: 'http://localhost:10090/efm/api/c2-protocol/heartbeat'
  nifi.c2.rest.url.ack: 'http://localhost:10090/efm/api/c2-protocol/acknowledge'
  nifi.c2.agent.heartbeat.period: '5000'
  nifi.c2.agent.identifier: '123-456-789'
  nifi.c2.agent.class: 'java-new'

In the mean time Ferenc Erdei is working on some updates on the MiNiFi side and was looking into passing the property values from there instead of having them in config.yml (which is not a great place for them). I remember vaguely that you mentioned that in the end the parameters shouldn't be in the config.yml but also that bootstrap.conf should go away (maybe I'm misremembering) but in that case what would be the best way to add these to nifi.properties?

**Pull request merged**